fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java

```diff
@@ -0,0 +1,371 @@
+
+        // build filter conjunct, get the alias of the agg output and extract its child.
+        // then replace the agg to window function, then build conjunct
+        NamedExpression aggOut = agg.getOutputExpressions().get(0);
+        Expression aggOutExpr = aggOut.child(0);
+        aggOutExpr = new FunctionReplacer().replace(aggOutExpr, windowFunctionAlias.toSlot());
+
+        windowFilterConjunct = ((ComparisonPredicate) windowFilterConjunct
+                .withChildren(windowFilterConjunct.child(0), aggOutExpr));
+
+        LogicalFilter newFilter = ((LogicalFilter) node.withChildren(apply.left()));
+        LogicalWindow newWindow = new LogicalWindow<>(ImmutableList.of(windowFunctionAlias), newFilter);
+        LogicalFilter windowFilter = new LogicalFilter<>(ImmutableSet.of(windowFilterConjunct), newWindow);
+        return windowFilter;
+    }
+
+    // check children's nodes because query process will be changed
+    private boolean checkPlanType() {
+        return outerPlans.stream().allMatch(p -> LEFT_SUPPORTED_PLAN.contains(p.getClass()))
+                && innerPlans.stream().allMatch(p -> RIGHT_SUPPORTED_PLAN.contains(p.getClass()));
+    }
+
+    // check aggregation of inner scope
+    private boolean checkAggType() {
+        List<LogicalAggregate> aggSet = innerPlans.stream().filter(LogicalAggregate.class::isInstance)
+                .map(LogicalAggregate.class::cast)
+                .collect(Collectors.toList());
+        if (aggSet.size() > 1) {
+            // window functions don't support nesting.
+            return false;
+        }
+        aggOp = aggSet.get(0);
+        functions = ((List<AggregateFunction>) ExpressionUtils.<AggregateFunction>collectAll(
+                aggOp.getOutputExpressions(), AggregateFunction.class::isInstance));
+        Preconditions.checkArgument(functions.size() == 1);
+        return functions.stream().allMatch(f -> SUPPORTED_FUNCTION.contains(f.getClass()) && !f.isDistinct());
+    }
+
+    // check if the relations of the outer's includes the inner's
+    private boolean checkRelation(List<Expression> correlatedSlots) {
+        List<LogicalRelation> outerTables = outerPlans.stream().filter(LogicalRelation.class::isInstance)
+                .map(LogicalRelation.class::cast)
+                .collect(Collectors.toList());
+        List<LogicalRelation> innerTables = innerPlans.stream().filter(LogicalRelation.class::isInstance)
+                .map(LogicalRelation.class::cast)
+                .collect(Collectors.toList());
+
+        Set<Long> outerIds = outerTables.stream().map(node -> node.getTable().getId()).collect(Collectors.toSet());
+        Set<Long> innerIds = innerTables.stream().map(node -> node.getTable().getId()).collect(Collectors.toSet());
+
+        if (outerTables.size() != outerIds.size() || innerTables.size() != innerIds.size()
+                || outerIds.size() != innerIds.size() + 1) {
```

Comment:
why we need them ? `outerTables.size() != outerIds.size() || innerTables.size() != innerIds.size()`